exchanges/stream/websocket_connection.go

```diff
@@ -180,6 +180,7 @@ func (w *WebsocketConnection) SetupPingHandler(handler PingHandler) {
 						"%v websocket connection: ping handler failed to send message [%s]",
 						w.ExchangeName,
 						handler.Message)
+					w.readMessageErrors <- err
```

Comment:
Agreed, If there is no routine waiting I would log the error as well in the default case.

Reply:
There are a few things to adress here:

First, if I remember correctlt, I have tried setting delay to a much shorter interval, I think one minute it is. The disconnection problem still exists. I tried trouble shooting this by adding a log message where
just before `w.Connection.WriteControl(...)`, like ` log.Info(..., "send pong")`, the result is some thing like

```
send pong
send pong
send pong
<after some hours, no more websocket updates and "send pong"s>
```

and that's why I doubt for some reason I got missed on some PING messages and failed to respond to them.

A note here: isn't the logic of GrorillaHandler that when a PING is received, a reponse PONG is sent back no later than `delay` duration after the PING is received? I tried to read through gorilla's documentation but sorry for my lack of knowledge I am quite confused about this.

For your second point, I raised an issue in an earlier issue that the websocket (centralized) reconnection mechanism has some bugs in Shutdown(). It deadlocks (I can consistently reproduce). So I throw it away and use Binance package as standalone, I have no other dependencies on other exchanges. Everything is fine in the standalone mode, except for this annonying disconnection problem which happens each a few hours.

Last point, I don't have any order book subscription needs right now, so I simply subscribed the order status stream, I throw away other messages in DataHandler.

Reply:
I am perfectly fine if `exchanges/stream` module can handle reconnection properly, but unfortunately it still has some issues. If the maintainers plan to fix the centralised websocket system soon I am happy to use them rather than my kind of dirty tricks.

Reply:
Interesting, thank you for responding in detail and testing! In regards to websocket freezing and failing to reconnect, we have this PR which should solve those problems: https://github.com/thrasher-corp/gocryptotrader/pull/754 once its merged 🎉 

Overall, the change from using gorilla websocket pinghandler to our own pinghandler doesn't have much of an impact and ensures that the pong message is sent for Binance, so I'm fine with it here.

You will need to make the change regarding `w.readMessageErrors` `default` case as above, as that can become one of the reasons a websocket cannot reconnect using our global system

Reply:
Update:

Hey guys, after some more experimenting, I found out I can make my code work WITHOUT this PR. I simply handle errors on the client side, doing things like this:

```golang
func (c *Coin) autoReconnect() {
	for {
		select {
		case err := <-c.websocket.ReadMessageErrors:
			log.Printf("detached read messaeg error: %v, reconnect and resubscribe", err)
			var attempt int
			for {
				attempt++
				err = c.binance.WsConnect()
				if err != nil {
					log.Printf("%v, retry in %s", err, reconnectDelay)
					if attempt > reconnectMaxAttempt {
						log.Printf("reconnect failed more than %d times", reconnectMaxAttempt)
						break
					}
				} else {
					log.Printf("reconnect ok")
					break
				}
				time.Sleep(reconnectDelay)
			}
			err = c.wsResubscribe()
			if err != nil {
				log.Printf("cannot resubscribe: %v", err)
			} else {
				log.Printf("resubscribe ok")
			}
		}
	}
}
```

It turns out I can get an error like "Abornal close..." when the connection is broken from `websocket.ReadMessageErrors`, so I put all the reconnection stuff in my client code.

I will close this PR, hoping to see #754 merged soon.